sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/FileSystemAPITest.groovy

```diff
@@ -946,6 +947,137 @@ class FileSystemAPITest extends APISpec {
     }
     // TODO (gapra): Add more get paths tests (Github issue created)
 
+    def "List deleted paths"() {
+        setup:
+        System.out.println(new PathDeletedItem("foo", false, "bar", OffsetDateTime.now(), 5).isPrefix())
+        enableSoftDelete()
+
+        def fc1 = fsc.getFileClient(generatePathName())
+        fc1.create(true)
+        fc1.delete()
+
+        when:
+        List<PathDeletedItem> deletedBlobs = fsc.listDeletedPaths().stream().collect(Collectors.toList())
+
+        then:
+        deletedBlobs.size() == 1
+        !deletedBlobs.get(0).isPrefix()
+        deletedBlobs.get(0).getName() == fc1.getFileName()
+        deletedBlobs.get(0).getDeletedOn() != null
+        deletedBlobs.get(0).getDeletionId() != null
+        deletedBlobs.get(0).getRemainingRetentionDays() != null
+
+        cleanup:
+        disableSoftDelete()
+    }
+
+    def "List deleted paths path"() {
+        setup:
+        enableSoftDelete()
+
+        def dir = fsc.getDirectoryClient(generatePathName())
+        dir.create()
+        def fc1 = dir.getFileClient(generatePathName()) // Create one file under the path
+        fc1.create(true)
+        fc1.delete()
+
+        def fc2 = fsc.getFileClient(generatePathName()) // Create another file not under the path
+        fc2.create()
+        fc2.delete()
+
+        when:
+        def deletedBlobs = fsc.listDeletedPaths(new ListDeletedPathsOptions().setPath(dir.getDirectoryName()), null)
+
+        then:
+        deletedBlobs.size() == 1
+        !deletedBlobs.first().isPrefix()
+        deletedBlobs.first().getName() == dir.getDirectoryName() + "/" + fc1.getFileName()
+
+        cleanup:
+        disableSoftDelete()
+    }
+
+    def "List deleted paths error"() {
+        setup:
+        enableSoftDelete()
+
+        fsc = primaryDataLakeServiceClient.getFileSystemClient(generateFileSystemName())
+
+        when:
+        fsc.listDeletedPaths().last()
+
+        then:
+        thrown(DataLakeStorageException)
+
+        cleanup:
+        disableSoftDelete()
+    }
+
+    def "Restore path"() {
+        setup:
+        enableSoftDelete()
+
+        def dir = fsc.getDirectoryClient(generatePathName())
+        dir.create()
+        dir.delete()
+
+        def deletionId = fsc.listDeletedPaths().first().getDeletionId()
+
+        when:
+        fsc.restorePath(dir.getDirectoryName(), deletionId)
+
+        then:
+        dir.getProperties() != null
+
+        cleanup:
+        disableSoftDelete()
+    }
+
+    @Unroll
+    def "Restore path special characters"() {
+        setup:
+        enableSoftDelete()
+
+        dirName = URLEncoder.encode(dirName)
+        def dir = fsc.getDirectoryClient(dirName)
+        dir.create()
+        dir.delete()
+
+        def deletionId = fsc.listDeletedPaths().first().getDeletionId()
+
+        when:
+        fsc.restorePath(URLEncoder.encode(dir.getDirectoryName()), deletionId)
```

Comment:
Should we do a getPathName and make sure its correct?

Reply:
Done